pkg/reconciler/apis/replicationclusterrolebinding/replicationclusterrolebinding_controller.go

```diff
@@ -0,0 +1,254 @@
+func (c *controller) Start(ctx context.Context, numThreads int) {
+	defer runtime.HandleCrash()
+	defer c.queue.ShutDown()
+
+	logger := logging.WithReconciler(klog.FromContext(ctx), ControllerName)
+	ctx = klog.NewContext(ctx, logger)
+	logger.Info("Starting controller")
+	defer logger.Info("Shutting down controller")
+
+	for i := 0; i < numThreads; i++ {
+		go wait.UntilWithContext(ctx, c.startWorker, time.Second)
+	}
+
+	<-ctx.Done()
+}
+
+func (c *controller) startWorker(ctx context.Context) {
+	for c.processNextWorkItem(ctx) {
+	}
+}
+
+func (c *controller) processNextWorkItem(ctx context.Context) bool {
+	// Wait until there is a new item in the working queue
+	k, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	key := k.(string)
+
+	logger := logging.WithQueueKey(klog.FromContext(ctx), key)
+	ctx = klog.NewContext(ctx, logger)
+	logger.V(4).Info("processing key")
+
+	// No matter what, tell the queue we're done with this key, to unblock
+	// other workers.
+	defer c.queue.Done(key)
+
+	if err := c.process(ctx, key); err != nil {
+		runtime.HandleError(fmt.Errorf("%q controller failed to sync %q, err: %w", ControllerName, key, err))
+		c.queue.AddRateLimited(key)
+		return true
+	}
+	c.queue.Forget(key)
+	return true
+}
+
+func (c *controller) process(ctx context.Context, key string) error {
+	cluster, _, name, err := kcpcache.SplitMetaClusterNamespaceKey(key)
+	if err != nil {
+		runtime.HandleError(err)
+		return nil
+	}
+
+	crb, err := c.clusterRoleBindingLister.Cluster(cluster).Get(name)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return nil // object deleted before we handled it
+		}
+		return err
+	}
+
+	logger := logging.WithObject(klog.FromContext(ctx), crb)
+	ctx = klog.NewContext(ctx, logger)
+
+	// is a maximum-permission-policy subject?
+	replicate := false
+	for _, s := range crb.Subjects {
+		if strings.HasPrefix(s.Name, apisv1alpha1.MaximalPermissionPolicyRBACUserGroupPrefix) && (s.Kind == rbacv1.UserKind || s.Kind == rbacv1.GroupKind) && s.APIGroup == rbacv1.GroupName {
+			replicate = true
+			break
+		}
+	}
+
+	// references relevant ClusterRole?
+	if !replicate && crb.RoleRef.Kind == "ClusterRole" && crb.RoleRef.APIGroup == rbacv1.GroupName {
+		cr, err := c.clusterRoleLister.Cluster(cluster).Get(crb.RoleRef.Name)
+		if err != nil && !errors.IsNotFound(err) {
+			return err
+		}
+		if cr != nil && replicationclusterrole.HasBindOrContentRule(cr) {
+			replicate = true
+		}
+	}
+
+	_, found := crb.Labels[ReplicateLabelKey]
+	if replicate == found {
+		return nil
+	}
+
+	patch := fmt.Sprintf(`{"metadata":{"labels":{"%s":null}}}`, ReplicateLabelKey)
+	if replicate {
+		patch = fmt.Sprintf(`{"metadata":{"labels":{"%s":"%t"}}}`, ReplicateLabelKey, replicate)
+	}
+
+	logger.V(4).Info("patching ClusterRoleBinding", "patch", patch)
+	_, err = c.kubeClusterClient.Cluster(cluster.Path()).RbacV1().ClusterRoleBindings().Patch(ctx, crb.Name, types.MergePatchType, []byte(patch), metav1.PatchOptions{})
```

Comment:
wondering if we could use `committer.NewCommitter`

Reply:
no, I can't. But I wrote one I can use. Stay tuned :)